db/db.go

```diff
@@ -0,0 +1,260 @@
+// nothing is found, returns empty list.
+func (m *mongo) ReadEntities(q query.Query) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	var entities []Entity
+	c := s.DB(m.database).C(m.collection)
+	err := c.Find(mgoQuery).All(&entities)
+	if err != nil {
+		return nil, ErrNotFound
+	}
+
+	return entities, nil
+}
+
+// UpdateEntities does a db.collection.findAndModify() operation
+// (https://docs.mongodb.com/manual/reference/method/db.collection.findAndModify/)
+// on all docs returned from query. Returns diff of entities (old values for
+// fields changed, in addition to including "_id" field/value) or an error, if
+// one occured.
+func (m *mongo) UpdateEntities(q query.Query, u Entity) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := s.DB(m.database).C(m.collection)
+
+	// List of IDs for docs to update
+	ids, err := idsForQuery(c, mgoQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change to make
+	change := mgo.Change{Update: bson.M{"$set": u}}
+
+	// Query for each document and apply update
+	var diffs []Entity
+	for _, id := range ids {
+		var diff Entity
+		_, err := c.Find(bson.M{"_id": id}).Select(selectMap(u)).Apply(change, &diff)
+		if err != nil {
+			return nil, ErrUpdateEntity
+		}
+
+		diffs = append(diffs, diff)
+	}
+
+	return diffs, nil
+}
+
+// Deletes many Entity objects from DB. Returns old entities that were removed
+// or an error, if one occured.
+func (m *mongo) DeleteEntities(q query.Query) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := s.DB(m.database).C(m.collection)
+
+	// List of IDs for docs to update
+	ids, err := idsForQuery(c, mgoQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change to make
+	change := mgo.Change{Remove: true}
+
+	// Query for each document and delete it
+	var results []Entity
+	for _, id := range ids {
+		var result Entity
+		_, err := c.Find(bson.M{"_id": id}).Apply(change, &result)
+		if err != nil {
+			return nil, ErrDeleteEntity
+		}
+
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+// Translates Query object to bson.M object, which will
+// be used to query DB.
+func translateQuery(q query.Query) bson.M {
+	mgoQuery := make(bson.M)
+	for _, p := range q.Predicates {
+		switch p.Operator {
+		case "exists":
+			mgoQuery[p.Label] = bson.M{"$exists": true}
+		case "!":
+			mgoQuery[p.Label] = bson.M{"$exists": false}
+		default:
+			mgoQuery[p.Label] = bson.M{operatorMap[p.Operator]: p.Value}
+		}
+	}
+	return mgoQuery
+}
+
+// Get all ids of docs that satisfy query
+func idsForQuery(c *mgo.Collection, q bson.M) ([]string, error) {
+	var resultIDs []map[string]string
+	err := c.Find(q).Select(bson.M{"_id": 1}).All(&resultIDs)
+	if err != nil {
+		return nil, ErrNotFound
+	}
+
+	var ids []string
+	for _, id := range resultIDs {
```

Comment:
Small optimization: since we know the length of `resultIDs`, we can avoid dynamically expanding allocation with `append` and instead do:
```go
ids := make(string, len(resultIDs))
for i, id := range resultIDs {
    ids[i] = id["_id"]
}
```